src/js/shield.js

```diff
@@ -6,36 +6,44 @@ import * as ShieldDef from "./shield_defs.js";
 import * as ShieldText from "./shield_text.js";
 import * as ShieldDraw from "./shield_canvas_draw.js";
 import * as Gfx from "./screen_gfx.js";
+import rgba from "color-rgba";
 
-function loadShield(ctx, shield, bannerCount, verticalReflect) {
-  var drawCtx = Gfx.getGfxContext(shield.data);
-  var imgData = drawCtx.createImageData(shield.data.width, shield.data.height);
+function loadPixel(source, dest, sourceOffset, destOffset, colorLighten) {
+  dest[destOffset] = Math.max(source[sourceOffset], colorLighten[0]); //Red
+  dest[destOffset + 1] = Math.max(source[sourceOffset + 1], colorLighten[1]); //Green
+  dest[destOffset + 2] = Math.max(source[sourceOffset + 2], colorLighten[2]); //Blue
+  dest[destOffset + 3] = source[sourceOffset + 3]; //Alpha
+}
 
-  for (var i = 0; i < shield.data.data.length; i++) {
-    imgData.data[i] = shield.data.data[i];
+function rgbaMatrix(colorLighten) {
+  if (typeof colorLighten !== "undefined") {
+    return rgba(colorLighten);
   }
+  return [0, 0, 0, 0];
+}
 
-  drawCtx.putImageData(imgData, 0, 0);
+function loadSprite(ctx, shield, bannerCount, verticalReflect, colorLighten) {
+  let imgData = ctx.createImageData(shield.data.width, shield.data.height);
+  let yOffset = bannerCount * ShieldDef.bannerSizeH + ShieldDef.topPadding;
+  let lighten = rgbaMatrix(colorLighten);
 
   if (verticalReflect == null) {
-    ctx.drawImage(
-      drawCtx.canvas,
-      0,
-      bannerCount * ShieldDef.bannerSizeH + ShieldDef.topPadding
-    );
+    for (let i = 0; i < shield.data.data.length; i += 4) {
+      loadPixel(shield.data.data, imgData.data, i, i, lighten);
+    }
   } else {
-    ctx.save();
-    ctx.scale(1, -1);
-    ctx.drawImage(
-      drawCtx.canvas,
-      0,
-      bannerCount * ShieldDef.bannerSizeH +
-        ShieldDef.topPadding -
-        drawCtx.canvas.height -
-        2 * ShieldDraw.PXR
-    );
-    ctx.restore();
+    //4 bytes/px, copy in reverse vertical order.
```

Comment:
This feels a little weird, you're reflecting the image vertically and also adjusting the color lightness at the same time.

Is it done in one operation for performance reasons? What is being reflected and why?

Reply:
At present, the vertical flipping is done for the Hawaii shield only, and I didn't change the existing logic. There's shields in Japan and Taiwan that are an upside-down version of this same shape. Since there's only one shield that uses this feature, we could certainly just make a graphic for it.

Since the lighten transformation is a pixel-twiddling operation on the background graphic, and we were already doing pixel copies, I figured I'd just add the logic at the time of copy. I applied the color lightness to both the flip and no-flip versions more or less just for completeness, since there aren't any shields that are flipped AND recolored.

While the code looks low level, it does seem to be performant. If we unwound the flipping capability it would simplify the code at the expense of one extra graphic (and the hope that we don't identify other shields that are upside-down variants of something.

![image](https://user-images.githubusercontent.com/3254090/215295032-44566bbb-99ea-4b4f-8a47-90d242a52c9d.png)